driver-core/src/main/com/mongodb/client/model/WindowedComputations.java

```diff
@@ -0,0 +1,769 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.client.model;
+
+import com.mongodb.annotations.Beta;
+import com.mongodb.client.model.Windows.Bound;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonDocument;
+import org.bson.BsonDocumentWriter;
+import org.bson.BsonType;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.mongodb.assertions.Assertions.isTrueArgument;
+import static com.mongodb.client.model.MongoTimeUnit.MONTH;
+import static com.mongodb.client.model.MongoTimeUnit.YEAR;
+import static org.bson.assertions.Assertions.notNull;
+
+/**
+ * Builders for {@linkplain WindowedComputation windowed computations} used in the
+ * {@link Aggregates#setWindowFields(Object, Bson, List) $setWindowFields} pipeline stage
+ * of an aggregation pipeline. Each windowed computation is a triple:
+ * <ul>
+ *     <li>A window function. Some functions require documents in a window to be sorted
+ *     (see {@code sortBy} in {@link Aggregates#setWindowFields(Object, Bson, List)}).</li>
+ *     <li>An optional {@linkplain Window window}, a.k.a. frame.
+ *     Specifying {@code null} window is equivalent to specifying an unbounded window,
+ *     i.e., a window with both ends specified as {@link Bound#UNBOUNDED}.
+ *     Some window functions require to specify an explicit unbounded window instead of specifying {@code null}.</li>
+ *     <li>A path to an output field to be computed by the window function over the window.</li>
+ * </ul>
+ * A windowed computation is similar to an {@linkplain Accumulators accumulator} but does not result in folding documents constituting
+ * the window into a single document.
+ *
+ * @mongodb.driver.manual meta/aggregation-quick-reference/#field-paths Field paths
+ * @since 4.3
+ */
+@Beta
+public final class WindowedComputations {
+    /**
+     * Builds a computation of the sum of the evaluation results of the {@code expression} over the {@code window}.
+     *
+     * @param path The output field path.
+     * @param expression The expression.
+     * @param window The window.
+     * @param <TExpression> The expression type.
+     * @return The constructed {@link WindowedComputation}.
+     * @mongodb.driver.dochub core/window-functions-sum $sum
+     * @mongodb.server.release 5.0
+     */
+    public static <TExpression> WindowedComputation sum(final String path, final TExpression expression, @Nullable final Window window) {
```

Comment:
It's strange to me that for all of these windowed computations, the window itself is optional.  Can you try to explain that in the Javadoc at the class level, i.e. what the computations mean when no window is provided?

Also, I wonder how common it will be to have a null window?  Would it be better to overload the methods instead of requiring a null third parameter (that will introduce conflicts with Accumulators if we did that)?

Reply:
> It's strange to me that for all of these windowed computations, the window itself is optional

This is with accordance to the MQL syntax: explicit window is optional for the majority of functions.

> Can you try to explain that in the Javadoc at the class level, i.e. what the computations mean when no window is provided?

Yes, the documentation of `WindowedComputations` states:

```
Specifying null window is equivalent to specifying an unbounded window,
i.e., a window with both ends specified as Windows.Bound.UNBOUNDED. 
Some window functions require to specify an explicit unbounded window instead of specifying null.
```

> Would it be better to overload the methods instead of requiring a null third parameter 

I am not sure this is actually useful, but since adding such overloads is a backward-compatible change, we could do it if there is an interest expressed by users. Our approach of duplicating documentation (including documentation of overloaded methods) instead of linking to a single place that could be used to document multiple program elements also results in an additional burden per each overload, and makes overload less favourable.

Reply:
Thanks for the explanation.  Resolving the thread.